processes/fulfillURLManifest.py

```diff
@@ -62,31 +62,37 @@ def update_manifest(self, metadataObject, bucketName, currKey):
 
         metadataJSON = json.loads(metadataObject['Body'].read().decode("utf-8"))
         metadataJSONCopy = copy.deepcopy(metadataJSON)
+
+        #Checking if record asscociated with manifest is public domain or in copyright
+        checkRightsStatus = 'in copyright'
+        checkRightsStatus = self.checkMetaData(metadataJSON, checkRightsStatus)
+
+        if checkRightsStatus == 'in copyright':
         
-        counter = 0
-    
-        metadataJSON, counter = self.linkFulfill(metadataJSON, counter)
-        metadataJSON, counter = self.readingOrderFulfill(metadataJSON, counter)
-        metadataJSON, counter = self.resourceFulfill(metadataJSON, counter)
-        metadataJSON, counter = self.tocFulfill(metadataJSON, counter)
-
-        if counter >= 4: 
-            for link in metadataJSON['links']:
-                self.fulfillFlagUpdate(link)
-
-        self.closeConnection()
-
-        if metadataJSON != metadataJSONCopy:
-            try:
-                fulfillManifest = json.dumps(metadataJSON, ensure_ascii = False)
-                return self.s3Client.put_object(
-                    Bucket=bucketName, 
-                    Key=currKey, 
-                    Body=fulfillManifest, ACL= 'public-read', 
-                    ContentType = 'application/json'
-                )
-            except ClientError as e:
-                logging.error(e)
+            counter = 0
+        
+            metadataJSON, counter = self.linkFulfill(metadataJSON, counter)
+            metadataJSON, counter = self.readingOrderFulfill(metadataJSON, counter)
+            metadataJSON, counter = self.resourceFulfill(metadataJSON, counter)
+            metadataJSON, counter = self.tocFulfill(metadataJSON, counter)
+
+            if counter >= 4: 
+                for link in metadataJSON['links']:
+                    self.fulfillFlagUpdate(link)
```

Comment:
The counter logic is unclear. What are your thoughts on refactoring this and the methods are also unclear?

What does it mean to fulfill the link, reading order, resource, toc, etc? What does the counter mean? Why is it if the counter >=4 that we update the fulfill flag?

Reply:
The counter is there to keep track and make sure all the necessary pdf/epub links in the manifest JSON document are replaced with fulfill endpoints. The structure of our manifest documents always have a links, readingOrder, resource, and toc dictionary that include pdf/epub links. Therefore to ensure that the fulfill_limited_access flag doesn't become true while one of the links is not replaced with a fulfill endpoint for some reason or another the counter must be at least 4 to ensure all four dictionaries have been updated properly.

I can refactor the counter to instead be four separate flags that become true for each fulfill method to make this more clear. What do you think?

Reply:
I think what you can do is try to use try/except and error handling for this. There are interesting edge cases here like what happens if there are 4 links in linkFulfill but no reading order, resource, nor toc links? 

I think we should try to do it without a flag and without a counter. One way to think about it is to simplify the functions that you are calling in this block. What if we get all the links ids for the links, reading order links, resource links, and toc links? 

We can use error handling to answer questions like what happens when there are no links found.

Reply:
For now I'll implement a try/except block to error handle those edge cases but I'll keep the counter since I have other tasks to focus on at the moment and simplifying the functions to one method isn't necessary at this time since the process is working at this moment.